sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -124,51 +122,90 @@ private MqttIO() {
+     * @return A connection configuration to the MQTT broker.
+     */
     public static ConnectionConfiguration create(String serverUri, String topic, String clientId) {
       checkArgument(serverUri != null,
-          "MqttIO.ConnectionConfiguration.create(serverUri, topic) called with null "
-              + "serverUri");
+          "MqttIO.ConnectionConfiguration.create(getServerUri, getTopic) called with null "
+              + "getServerUri");
       checkArgument(topic != null,
-          "MqttIO.ConnectionConfiguration.create(serverUri, topic) called with null "
-              + "topic");
-      checkArgument(clientId != null, "MqttIO.ConnectionConfiguration.create(serverUri, topic, "
-          + "clientId) called with null clientId");
-      return new AutoValue_MqttIO_ConnectionConfiguration(serverUri, topic, clientId);
+          "MqttIO.ConnectionConfiguration.create(getServerUri, getTopic) called with null "
+              + "getTopic");
+      checkArgument(clientId != null, "MqttIO.ConnectionConfiguration.create(getServerUri,"
+          + "getTopic, getClientId) called with null getClientId");
+      return new AutoValue_MqttIO_ConnectionConfiguration.Builder().setServerUri(serverUri)
+          .setTopic(topic).setClientId(clientId).build();
     }
 
-    private void populateDisplayData(DisplayData.Builder builder) {
-      builder.add(DisplayData.item("serverUri", serverUri()));
-      builder.add(DisplayData.item("topic", topic()));
-      builder.add(DisplayData.item("clientId", clientId()));
+    public ConnectionConfiguration withUsername(String username) {
+      return builder().setUsername(username).build();
     }
 
-    private MqttClient getClient(boolean random) throws MqttException {
-      String id = clientId();
-      if (random) {
-        id = clientId() + "-" + MqttClient.generateClientId();
-      }
-      MqttClient client = new MqttClient(serverUri(), id);
-      client.connect();
-      return client;
+    public ConnectionConfiguration withPassword(String password) {
+      return builder().setPassword(password).build();
     }
 
-    private MqttClient getClient() throws MqttException {
-      return getClient(false);
+    private void populateDisplayData(DisplayData.Builder builder) {
+      builder.add(DisplayData.item("serverUri", getServerUri()));
+      builder.add(DisplayData.item("topic", getTopic()));
+      builder.addIfNotNull(DisplayData.item("clientId", getClientId()));
+      builder.addIfNotNull(DisplayData.item("username", getUsername()));
+    }
+
+    private MQTT getClient() throws Exception {
+      MQTT client = new MQTT();
+      client.setHost(getServerUri());
+      if (getUsername() != null) {
+        client.setUserName(getUsername());
+        client.setPassword(getPassword());
+      }
+      if (getClientId() != null) {
+        client.setClientId(getClientId() + "-" + System.currentTimeMillis());
```

Comment:
System.currentTimeMillis() is not unique. Please use an actual unique ID generator.